crypto/crmf/crmf_lib.c

```diff
@@ -505,10 +508,11 @@ int OSSL_CRMF_MSGS_verify_popo(const OSSL_CRMF_MSGS *reqs,
             it = ASN1_ITEM_rptr(OSSL_CRMF_CERTREQUEST);
             asn = req->certReq;
         }
-        if (ASN1_item_verify_with_libctx(it, sig->algorithmIdentifier,
-                                         sig->signature, asn, NULL,
-                                         X509_PUBKEY_get0(pubkey),
-                                         libctx, propq) < 1)
+        if ((pkey = X509_PUBKEY_get0(pubkey)) == NULL
+            || !evp_pkey_downgrade(pkey)
```

Comment:
Why? Is this still necessary now that #11808 has gone in?

Reply:
I'll check

Reply:
Removing this change resulted in:

``` console
$ make test TESTS=65
make depend && make _tests
make[1]: Entering directory '/home/levitte/gitwrk/openssl.net/official/_levitte-ossl_store-for-providers'
make[1]: Leaving directory '/home/levitte/gitwrk/openssl.net/official/_levitte-ossl_store-for-providers'
make[1]: Entering directory '/home/levitte/gitwrk/openssl.net/official/_levitte-ossl_store-for-providers'
( SRCTOP=../levitte-ossl_store-for-providers \
  BLDTOP=. \
  PERL="/usr/bin/perl" \
  FIPSKEY="f4556650ac31d35461610bac4ed81b1a181b2d8a43ea2854cbae22ca74560813" \
  EXE_EXT= \
  /usr/bin/perl ../levitte-ossl_store-for-providers/test/run_tests.pl 65 )
Use of uninitialized value $ENV{"HARNESS_JOBS"} in numeric gt (>) at ../levitte-ossl_store-for-providers/test/run_tests.pl line 60.
65-test_cmp_asn.t ...... ok   
65-test_cmp_client.t ... 
AddressSanitizer:DEADLYSIGNAL
=================================================================
==10304==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000000 (pc 0x560d940b934b bp 0x7ffe8a1584d0 sp 0x7ffe8a158400 T0)
==10304==The signal is caused by a READ memory access.
==10304==Hint: address points to the zero page.
    #0 0x560d940b934a in ASN1_item_verify_ctx ../levitte-ossl_store-for-providers/crypto/asn1/a_verify.c:166
    #1 0x560d940b8d78 in ASN1_item_verify_with_libctx ../levitte-ossl_store-for-providers/crypto/asn1/a_verify.c:106
    #2 0x560d93ee0b30 in OSSL_CRMF_MSGS_verify_popo ../levitte-ossl_store-for-providers/crypto/crmf/crmf_lib.c:510
    #3 0x560d93edd78b in ossl_cmp_verify_popo ../levitte-ossl_store-for-providers/crypto/cmp/cmp_vfy.c:826
    #4 0x560d93ed5972 in process_cert_request ../levitte-ossl_store-for-providers/crypto/cmp/cmp_server.c:209
    #5 0x560d93ed764f in OSSL_CMP_SRV_process_request ../levitte-ossl_store-for-providers/crypto/cmp/cmp_server.c:532
    #6 0x560d93ed7e96 in OSSL_CMP_CTX_server_perform ../levitte-ossl_store-for-providers/crypto/cmp/cmp_server.c:645
    #7 0x560d93ec4437 in send_receive_check ../levitte-ossl_store-for-providers/crypto/cmp/cmp_client.c:174
    #8 0x560d93ec7081 in OSSL_CMP_exec_certreq ../levitte-ossl_store-for-providers/crypto/cmp/cmp_client.c:716
    #9 0x560d93e7f650 in execute_exec_certrequest_ses_test ../levitte-ossl_store-for-providers/test/cmp_client_test.c:113
    #10 0x560d93e7ff65 in test_exec_CR_ses ../levitte-ossl_store-for-providers/test/cmp_client_test.c:200
    #11 0x560d93e87270 in run_tests ../levitte-ossl_store-for-providers/test/testutil/driver.c:321
    #12 0x560d93e87cea in main ../levitte-ossl_store-for-providers/test/testutil/main.c:30
    #13 0x7fbcb6488e0a in __libc_start_main ../csu/libc-start.c:308
    #14 0x560d93e7d029 in _start (/home/levitte/gitwrk/openssl.net/official/_levitte-ossl_store-for-providers/test/cmp_client_test+0x164029)
AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV ../levitte-ossl_store-for-providers/crypto/asn1/a_verify.c:166 in ASN1_item_verify_ctx
==10304==ABORTING
../../util/wrap.pl ../../test/cmp_client_test ../../../levitte-ossl_store-for-providers/test/recipes/65-test_cmp_client_data/server.key ../../../levitte-ossl_store-for-providers/test/recipes/65-test_cmp_client_data/server.crt ../../../levitte-ossl_store-for-providers/test/recipes/65-test_cmp_client_data/client.key ../../../levitte-ossl_store-for-providers/test/recipes/65-test_cmp_client_data/client.crt ../../../levitte-ossl_store-for-providers/test/recipes/65-test_cmp_client_data/client.csr none => 1
not ok 1
65-test_cmp_client.t ... 1/? ---------------------------------------------------
AddressSanitizer:DEADLYSIGNAL
=================================================================
==10385==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000000 (pc 0x55fd4c58434b bp 0x7fffc637fa50 sp 0x7fffc637f980 T0)
==10385==The signal is caused by a READ memory access.
==10385==Hint: address points to the zero page.
    #0 0x55fd4c58434a in ASN1_item_verify_ctx ../levitte-ossl_store-for-providers/crypto/asn1/a_verify.c:166
    #1 0x55fd4c583d78 in ASN1_item_verify_with_libctx ../levitte-ossl_store-for-providers/crypto/asn1/a_verify.c:106
    #2 0x55fd4c3abb30 in OSSL_CRMF_MSGS_verify_popo ../levitte-ossl_store-for-providers/crypto/crmf/crmf_lib.c:510
    #3 0x55fd4c3a878b in ossl_cmp_verify_popo ../levitte-ossl_store-for-providers/crypto/cmp/cmp_vfy.c:826
    #4 0x55fd4c3a0972 in process_cert_request ../levitte-ossl_store-for-providers/crypto/cmp/cmp_server.c:209
    #5 0x55fd4c3a264f in OSSL_CMP_SRV_process_request ../levitte-ossl_store-for-providers/crypto/cmp/cmp_server.c:532
    #6 0x55fd4c3a2e96 in OSSL_CMP_CTX_server_perform ../levitte-ossl_store-for-providers/crypto/cmp/cmp_server.c:645
    #7 0x55fd4c38f437 in send_receive_check ../levitte-ossl_store-for-providers/crypto/cmp/cmp_client.c:174
    #8 0x55fd4c392081 in OSSL_CMP_exec_certreq ../levitte-ossl_store-for-providers/crypto/cmp/cmp_client.c:716
    #9 0x55fd4c34a650 in execute_exec_certrequest_ses_test ../levitte-ossl_store-for-providers/test/cmp_client_test.c:113
    #10 0x55fd4c34af65 in test_exec_CR_ses ../levitte-ossl_store-for-providers/test/cmp_client_test.c:200
    #11 0x55fd4c352270 in run_tests ../levitte-ossl_store-for-providers/test/testutil/driver.c:321
    #12 0x55fd4c352cea in main ../levitte-ossl_store-for-providers/test/testutil/main.c:30
    #13 0x7f68faa97e0a in __libc_start_main ../csu/libc-start.c:308
    #14 0x55fd4c348029 in _start (/home/levitte/gitwrk/openssl.net/official/_levitte-ossl_store-for-providers/test/cmp_client_test+0x164029)
AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV ../levitte-ossl_store-for-providers/crypto/asn1/a_verify.c:166 in ASN1_item_verify_ctx
==10385==ABORTING
../../util/wrap.pl ../../test/cmp_client_test ../../../levitte-ossl_store-for-providers/test/recipes/65-test_cmp_client_data/server.key ../../../levitte-ossl_store-for-providers/test/recipes/65-test_cmp_client_data/server.crt ../../../levitte-ossl_store-for-providers/test/recipes/65-test_cmp_client_data/client.key ../../../levitte-ossl_store-for-providers/test/recipes/65-test_cmp_client_data/client.crt ../../../levitte-ossl_store-for-providers/test/recipes/65-test_cmp_client_data/client.csr default ../../../levitte-ossl_store-for-providers/test/default.cnf => 1
not ok 2
# ------------------------------------------------------------------------------
AddressSanitizer:DEADLYSIGNAL
=================================================================
==10496==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000000 (pc 0x55c53f10034b bp 0x7ffd8a1d8530 sp 0x7ffd8a1d8460 T0)
==10496==The signal is caused by a READ memory access.
==10496==Hint: address points to the zero page.
    #0 0x55c53f10034a in ASN1_item_verify_ctx ../levitte-ossl_store-for-providers/crypto/asn1/a_verify.c:166
    #1 0x55c53f0ffd78 in ASN1_item_verify_with_libctx ../levitte-ossl_store-for-providers/crypto/asn1/a_verify.c:106
    #2 0x55c53ef27b30 in OSSL_CRMF_MSGS_verify_popo ../levitte-ossl_store-for-providers/crypto/crmf/crmf_lib.c:510
    #3 0x55c53ef2478b in ossl_cmp_verify_popo ../levitte-ossl_store-for-providers/crypto/cmp/cmp_vfy.c:826
    #4 0x55c53ef1c972 in process_cert_request ../levitte-ossl_store-for-providers/crypto/cmp/cmp_server.c:209
    #5 0x55c53ef1e64f in OSSL_CMP_SRV_process_request ../levitte-ossl_store-for-providers/crypto/cmp/cmp_server.c:532
    #6 0x55c53ef1ee96 in OSSL_CMP_CTX_server_perform ../levitte-ossl_store-for-providers/crypto/cmp/cmp_server.c:645
    #7 0x55c53ef0b437 in send_receive_check ../levitte-ossl_store-for-providers/crypto/cmp/cmp_client.c:174
    #8 0x55c53ef0e081 in OSSL_CMP_exec_certreq ../levitte-ossl_store-for-providers/crypto/cmp/cmp_client.c:716
    #9 0x55c53eec6650 in execute_exec_certrequest_ses_test ../levitte-ossl_store-for-providers/test/cmp_client_test.c:113
    #10 0x55c53eec6f65 in test_exec_CR_ses ../levitte-ossl_store-for-providers/test/cmp_client_test.c:200
    #11 0x55c53eece270 in run_tests ../levitte-ossl_store-for-providers/test/testutil/driver.c:321
    #12 0x55c53eececea in main ../levitte-ossl_store-for-providers/test/testutil/main.c:30
    #13 0x7f4877536e0a in __libc_start_main ../csu/libc-start.c:308
    #14 0x55c53eec4029 in _start (/home/levitte/gitwrk/openssl.net/official/_levitte-ossl_store-for-providers/test/cmp_client_test+0x164029)
AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV ../levitte-ossl_store-for-providers/crypto/asn1/a_verify.c:166 in ASN1_item_verify_ctx
==10496==ABORTING
../../util/wrap.pl ../../test/cmp_client_test ../../../levitte-ossl_store-for-providers/test/recipes/65-test_cmp_client_data/server.key ../../../levitte-ossl_store-for-providers/test/recipes/65-test_cmp_client_data/server.crt ../../../levitte-ossl_store-for-providers/test/recipes/65-test_cmp_client_data/client.key ../../../levitte-ossl_store-for-providers/test/recipes/65-test_cmp_client_data/client.crt ../../../levitte-ossl_store-for-providers/test/recipes/65-test_cmp_client_data/client.csr fips ../../../levitte-ossl_store-for-providers/test/fips.cnf => 1
not ok 4
65-test_cmp_client.t ... 4/? ---------------------------------------------------
#   Failed test at ../levitte-ossl_store-for-providers/test/recipes/65-test_cmp_client.t line 47.
65-test_cmp_client.t ... Dubious, test returned 3 (wstat 768, 0x300)
Failed 3/4 subtests 
65-test_cmp_ctx.t ...... ok   
65-test_cmp_hdr.t ...... ok   
65-test_cmp_msg.t ...... ok   
65-test_cmp_protect.t .. ok   
65-test_cmp_server.t ... ok   
65-test_cmp_status.t ... ok   
65-test_cmp_vfy.t ...... ok   

Test Summary Report
-------------------
65-test_cmp_client.t (Wstat: 768 Tests: 4 Failed: 3)
  Failed tests:  1-2, 4
  Non-zero exit status: 3
Files=9, Tests=24,  7 wallclock secs ( 0.08 usr  0.02 sys +  3.56 cusr  0.59 csys =  4.25 CPU)
Result: FAIL
make[1]: *** [Makefile:3198: _tests] Error 1
make[1]: Leaving directory '/home/levitte/gitwrk/openssl.net/official/_levitte-ossl_store-for-providers'
make: *** [Makefile:3195: tests] Fel 2
```

With this change re-applied:

``` console
$ make test TESTS=65
make depend && make _tests
make[1]: Entering directory '/home/levitte/gitwrk/openssl.net/official/_levitte-ossl_store-for-providers'
make[1]: Leaving directory '/home/levitte/gitwrk/openssl.net/official/_levitte-ossl_store-for-providers'
make[1]: Entering directory '/home/levitte/gitwrk/openssl.net/official/_levitte-ossl_store-for-providers'
( SRCTOP=../levitte-ossl_store-for-providers \
  BLDTOP=. \
  PERL="/usr/bin/perl" \
  FIPSKEY="f4556650ac31d35461610bac4ed81b1a181b2d8a43ea2854cbae22ca74560813" \
  EXE_EXT= \
  /usr/bin/perl ../levitte-ossl_store-for-providers/test/run_tests.pl 65 )
Use of uninitialized value $ENV{"HARNESS_JOBS"} in numeric gt (>) at ../levitte-ossl_store-for-providers/test/run_tests.pl line 60.
65-test_cmp_asn.t ...... ok   
65-test_cmp_client.t ... ok   
65-test_cmp_ctx.t ...... ok   
65-test_cmp_hdr.t ...... ok   
65-test_cmp_msg.t ...... ok   
65-test_cmp_protect.t .. ok   
65-test_cmp_server.t ... ok   
65-test_cmp_status.t ... ok   
65-test_cmp_vfy.t ...... ok   
All tests successful.
Files=9, Tests=24,  8 wallclock secs ( 0.09 usr  0.01 sys +  3.44 cusr  0.56 csys =  4.10 CPU)
Result: PASS
make[1]: Leaving directory '/home/levitte/gitwrk/openssl.net/official/_levitte-ossl_store-for-providers'
```

Reply:
A better fix is:

```` diff
diff --git a/crypto/asn1/a_verify.c b/crypto/asn1/a_verify.c
index 2b2c46a854..e3471c8141 100644
--- a/crypto/asn1/a_verify.c
+++ b/crypto/asn1/a_verify.c
@@ -163,7 +163,7 @@ int ASN1_item_verify_ctx(const ASN1_ITEM *it, const X509_ALGOR *alg,
         }
 
         /* Check public key OID matches public key type */
-        if (EVP_PKEY_type(pknid) != pkey->ameth->pkey_id) {
+        if (!EVP_PKEY_is_a(pkey, OBJ_nid2sn(pknid))) {
             ASN1err(0, ASN1_R_WRONG_PUBLIC_KEY_TYPE);
             goto err;
         }
````

Reply:
Ah, yeah that helped